File: csharp/palindrome-products/PalindromeProducts.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class PalindromeProducts
{
    private enum CompareMode { Min, Max }

    private static (int, IEnumerable<(int, int)>) GetExtremePalindrome(int minFactor, int maxFactor, CompareMode mode)
    {
        var extremeFactor = mode == CompareMode.Max ? int.MinValue : int.MaxValue;
        var curExtrema = (extremeFactor: extremeFactor, factors: new List<(int, int)>());
        var hasFoundPalindrome = false;
        int curProduct;

        for (int i = minFactor; i <= maxFactor; i++)
        {
            for (int j = i; j <= maxFactor; j++)
            {
                curProduct = i * j;
                if (IsPalindrome(curProduct))
                {
                    hasFoundPalindrome = true;

                    if ((mode == CompareMode.Max && curProduct > curExtrema.extremeFactor) ||
                        (mode == CompareMode.Min && curProduct < curExtrema.extremeFactor))
                    {
                        curExtrema = (curProduct, new List<(int, int)> { (i, j) });
                    }
                    else if (curProduct == curExtrema.extremeFactor)
                    {
                        curExtrema.factors.Add((i, j));
                    }
                }
            }
        }

        if (hasFoundPalindrome)
        {
            return curExtrema;
        }
        else
        {
            throw new ArgumentException();
        }
    }

    public static bool IsPalindrome(int number)
    {
        var temp = number;
        var reversed = 0;
        int remaninder;

        while (temp > 0)
        {
            remaninder = temp % 10;
            reversed = reversed * 10 + remaninder;
            temp /= 10;
        }

        return number == reversed;
    }

    // public static bool IsPalindromeSlow(int number) =>
    //     number.ToString() == string.Concat(number.ToString().Reverse());

    public static (int, IEnumerable<(int, int)>) Largest(int minFactor, int maxFactor) =>
        GetExtremePalindrome(minFactor, maxFactor, CompareMode.Max);

    public static (int, IEnumerable<(int, int)>) Smallest(int minFactor, int maxFactor) =>
        GetExtremePalindrome(minFactor, maxFactor, CompareMode.Min);
}

File: ruby/palindrome-products/palindrome_products.rb
class Palindromes
  attr_reader :smallest, :largest

  Result = Struct.new(:value, :factors)

  def initialize(max_factor:, min_factor: 1)
    @min = min_factor
    @max = max_factor
  end

  def generate
    smallest_product = @max**2
    largest_product = 1

    (@min..@max).each do |a|
      (a..@max).each do |b|
        product = a * b

        next if product.to_s != product.to_s.reverse

        if product < smallest_product
          @smallest = Result.new(product, [[a, b]])
          smallest_product = product
        elsif product == smallest_product
          @smallest&.factors&.push([a, b])
        end

        if product > largest_product
          @largest = Result.new(product, [[a, b]])
          largest_product = product
        elsif product == largest_product
          @largest&.factors&.push([a, b])
        end
      end
    end
  end
end
